Auto instrument winston

garbados opened the pull request:
<!--
Thank you for submitting a Pull Request.

This code is leveraged to monitor critical services. Please consider the following:
* Tests are required.
* Performance matters.
* Features that are specific to just your app are unlikely to make it in.

Please fill out the relevant sections as follows:
* Proposed Release Notes: Bulleted list of recommended release notes for the change(s).
* Links: Any relevant links for the change.
* Details: In-depth description of changes, other technical notes, etc.
-->

## Proposed Release Notes

* Include Winston log enricher in default instrumentation, so logs are enriched automatically.

## Links

* https://github.com/newrelic/node-newrelic/issues/1142

## Details

This PR intercepts requiring Winston and modifies its logger factory, `createLogger`, to combine top-level and transport-level formatters with [our instrumented formatter](https://github.com/newrelic/newrelic-node-log-extensions/tree/main/packages/winston-log-enricher). This is done in such a way that user-supplied formatters modify logs before our formatter, which is applied last.

bizob2828 commented:
The reason the tests are failing in node 12 and 14 is because of peer deps.  `@newrelic/winston-enricher` requires winston to be a peer.  so you'd have to install winston at root of agent which we don't want to do.  I think a solution could be to refactor the `@newrelic/winston-enricher` to take an instance of winston at initialization(dependency injection) or require it.  If that becomes too messy we'll have to go back to drawing board and absract away common bits of log enricher and add to the auto instrumentation in agent.

bizob2828 commented:
Yea so this will work 

```
diff --git a/packages/winston-log-enricher/lib/createFormatter.js b/packages/winston-log-enricher/lib/createFormatter.js
index 040a9d0..ae1de42 100644
--- a/packages/winston-log-enricher/lib/createFormatter.js
+++ b/packages/winston-log-enricher/lib/createFormatter.js
@@ -4,10 +4,13 @@
  */
 
 'use strict'
-const winston = require('winston')
 const truncate = require('./truncate')
 
-module.exports = function createFormatter(newrelic) {
+module.exports = function createFormatter(newrelic, winston) {
+  // Allow for dependency injection in the case where the agent
+  // auto instruments winston
+  winston = winston || require('winston')
+

```

But will require an additional PR and release of the log enricher before this PR can be landed

bizob2828 commented:
I opened this PR that addressed the underlying issue https://github.com/newrelic/newrelic-node-log-extensions/pull/33

garbados commented:
I've updated this PR to account for the new config values added in https://github.com/newrelic/node-newrelic/commit/c8cdfba0d64a1326d57760ebade18e2566573c06 and for the Winston log enricher's new function signature added in https://github.com/newrelic/newrelic-node-log-extensions/commit/8f02672697e61c360244edf2e60b5c63aafce92f

Because the log enricher's new function signature hasn't been published yet, this PR currently fails CI. It will pass CI once the new version of the Winston log enricher has been published, and we have updated our package.json file with the new dependency version.

jordigh commented on test/versioned/winston/winston.tap.js:
```diff
@@ -0,0 +1,218 @@
+
+  const logStuff = (logger, streams) => {
+    // Log some stuff, both in and out of a transaction
+    logger.info('out of trans')
```
Since we do some things with log levels, although admittedly not in this test, should we also do some testing at various levels? For example, that we really don't log anything at the `debug` level because the logger was configured at the `info` level. So test that it really is so?

garbados replied:
I would be inclined to include those tests in the log enricher itself, like [here](https://github.com/newrelic/newrelic-node-log-extensions/blob/main/packages/winston-log-enricher/tests/versioned/winston.tap.js) rather than in this auto-instrumentation, where we are really just checking that we're able to hook onto the module loading to apply the log enricher. What do you think?

bizob2828 replied:
Good point. I'd make a case we should pare down these tests to avoid duplication in log enricher. But just enough to provide the log enricher can be instrumented and works as designed. Basically, how I found the problem with incorrect Winston instance getting enriched.

garbados replied:
I've updated the test suite to focus on the specific code paths in our instrumentation, rather than the formatter.

jordigh commented:
The PR looks good to me, but the fundamental idea of checking the config to see what's there and injecting our own config depending on what's there seems a bit off. Seems like it's brittle. Unhelpfully, I can't see what we could do instead that wouldn't be more brittle, like trying to inject into unstable internal winston API.

bizob2828 commented on lib/instrumentation/winston.js:
```diff
@@ -0,0 +1,46 @@
+
+module.exports = function instrument(agent, winston, _, shim) {
+  const config = agent.config
+  if (!config.application_logging || !config.application_logging.enabled) {
```
you also need to include `config.application_logging.local_decorating.enabled` I would also wrap this perhaps in some util function check as we will also need it in pino. You can also do all the checks and wrap them in parens and just do `!(<conditions>)`

`if (!(config.application_logging && config.application_logging.enabled && config.application_logging.local_decorating && config.application_logging.local_decorating.enabled))`

bizob2828 commented:
> The PR looks good to me, but the fundamental idea of checking the config to see what's there and injecting our own config depending on what's there seems a bit off. Seems like it's brittle. Unhelpfully, I can't see what we could do instead that wouldn't be more brittle, like trying to inject into unstable internal winston API.

Agreed.  We have a few use cases where we want to conditionally register instrumentation.  That's the intention of https://github.com/newrelic/node-newrelic/issues/1116, which we should expand to be generic.

garbados commented:
> the fundamental idea of checking the config to see what's there and injecting our own config depending on what's there seems a bit off

I'm not sure I understand. The only place this PR modifies the base config is during testing, where we do that to enable or disable logging to ensure our config checks catch properly. Am I missing something or otherwise misunderstanding?

jordigh commented:
> I'm not sure I understand. The only place this PR modifies the base config is during testing, where we do that to enable or disable logging to ensure our config checks catch properly. Am I missing something or otherwise misunderstanding?

[I mean this](https://github.com/newrelic/node-newrelic/pull/1146/files#diff-337511d015531c48ec1f9828260bb0eddaa40eb5a13523434a6be454f5b91a40R24), where we look at the existing config that our customer wrote and conditionally decide how to modify it. Is that not what we're doing? I'm very slightly concerned that the config could look different than what we're looking for here and our instrumentation wouldn't work.

garbados commented:
> conditionally decide how to modify it

The code you linked only wraps a method. It doesn't modify the user's config. Do you mean that we modify the options used to initialize Winston? That's standard practice for instrumentation, AFAIK.

bizob2828 commented on lib/instrumentation/winston.js:
```diff
@@ -0,0 +1,53 @@
+    )
+  ) {
+    shim.logger.debug('Application logging not enabled. Not instrumenting winston...')
+    return // do not instrument logs if logging is not enabled
```
nit: but that debug msg directly above states this

garbados replied:
👍

bizob2828 commented on lib/instrumentation/winston.js:
```diff
@@ -0,0 +1,53 @@
+
+  // wrap logger creation to combine or insert our formatter
+  shim.wrap(winston, 'createLogger', function wrapCreate(shim, createLogger) {
+    return function createWrappedLogger(opts = {}) {
```
this is probably ok but we typically just cast the arguments to an array and add more defense and call `.apply` and bind the proper this context to the original function

```js
    return function createWrappedLogger() {
      const args = shim.argsToArray.apply(shim, arguments)
      if (!shim.isObject(args[0])) {
        return createLogger.apply(this, args)
      }

      // combine top-level formatter with our own
      if ('format' in args[0]) {
        args[0].format = winston.format.combine(args[0].format, instrumentedFormatter())
      } else {
        args[0].format = instrumentedFormatter()
      }

      // combine transport-level formatters with our own
      if ('transports' in args[0]) {
        args[0].transports = args[0].transports.map((transport) => {
          if (transport.format) {
            transport.format = winston.format.combine(transport.format, instrumentedFormatter())
          }
          return transport
        })
      }

      return createLogger.apply(this, args)
    }
```

garbados replied:
Fixed to match your suggestion.

bizob2828 commented on lib/instrumentation/winston.js:
```diff
@@ -0,0 +1,53 @@
+      // combine top-level formatter with our own
+      if ('format' in opts) {
+        opts.format = winston.format.combine(opts.format, instrumentedFormatter())
+      } else {
```
you may be more familiar than I but shouldn't we check for transports in this if/else statement as well? it seems like we're adding a formatter always if we didn't specify a format.  So in the case where you define a transport we are adding a top level format with just the winston log enricher and then within the stream if you're specifying a format but not when you aren't.  Maybe we can talk through this case IRL

garbados replied:
I've adjusted the code to address this concern. Thanks for raising it!

garbados commented on lib/instrumentation/winston.js:
```diff
@@ -0,0 +1,53 @@
+
+module.exports = function instrument(agent, winston, _, shim) {
+  const config = agent.config
+  if (
```
👍

bizob2828 reviewed (commented):
Based on our planning convos we're going to keep this in PR until we can complete the rest of the work.  However, I did have a comment around some config checking that need tweaked

bizob2828 commented on lib/instrumentation/winston.js:
```diff
@@ -12,16 +12,20 @@ module.exports = function instrument(agent, winston, _, shim) {
   }
 
+  // set instrumentation variable so the enricher can know if it has been used twice
+  winston.__NR_instrumented = true
```
this happens automatically in `lib/shimmer` after this function gets called

bizob2828 commented on lib/instrumentation/winston.js:
```diff
@@ -12,16 +12,20 @@ module.exports = function instrument(agent, winston, _, shim) {
+        config.application_logging.local_decorating.enabled)
     )
   ) {
+    winston.__NR_instrumented = false
```
this will get overwritten by `lib/shimmer`. it sets it to true since we've registered instrumentation.  I think you're thinking of the right kind of thing.  I'm thinking though we may want to add something to the api that can identify it's coming from agent vs log enricher

bizob2828 replied:
we might be able to check for the existence of `config.application_logging && config.application_logging.enabled === false`

garbados commented:
Now that the latest Winston log enricher has been released, this PR should be ready for approval and merging. Note that we cannot merge https://github.com/newrelic/docs-website/pull/7458 until this is merged.

bizob2828 commented on test/versioned/winston/winston.tap.js:
```diff
@@ -0,0 +1,192 @@
+      })
+    })
+
+    const msgAssertFn = (t, msg) => {
```
when local_decorating is enabled it adds NR-LINKING metadata not keys to message.

bizob2828 reviewed (changes requested):
the versioned tests seem duplicative of the log enricher. i can fix these though since we're trying to land this today
